src/java/org/apache/cassandra/index/sai/utils/RangeUnionIterator.java

```diff
@@ -0,0 +1,166 @@
+                    continue;
+                if (candidate == null)
+                {
+                    candidate = range.peek();
+                    candidates.add(range);
+                }
+                else
+                {
+                    int cmp = candidate.compareTo(range.peek());
+                    if (cmp == 0)
+                        candidates.add(range);
+                    else if (cmp > 0)
+                    {
+                        candidates.clear();
+                        candidate = range.peek();
+                        candidates.add(range);
+                    }
+                }
+            }
+        }
+        if (candidates.isEmpty())
+            return endOfData();
+        candidates.forEach(RangeIterator::next);
+        return candidate;
+    }
+
+    protected void performSkipTo(PrimaryKey nextKey)
+    {
+        for (RangeIterator range : ranges)
+        {
+            if (range.hasNext())
+                range.skipTo(nextKey);
+        }
+    }
+
+    public void close() throws IOException
+    {
+        // Due to lazy key fetching, we cannot close iterator immediately
+        toRelease.forEach(FileUtils::closeQuietly);
+        ranges.forEach(FileUtils::closeQuietly);
+    }
+
+    public static Builder builder()
+    {
+        return new Builder();
+    }
+
+    public static RangeIterator build(List<RangeIterator> tokens)
+    {
+        return new Builder(tokens.size()).add(tokens).build();
+    }
+
+    public static class Builder extends RangeIterator.Builder
+    {
+        protected List<RangeIterator> rangeIterators;
+
+        public Builder()
+        {
+            super(IteratorType.UNION);
+            this.rangeIterators = new ArrayList<>();
+        }
+
+        public Builder(int size)
+        {
+            super(IteratorType.UNION);
+            this.rangeIterators = new ArrayList<>(size);
+        }
+
+        public RangeIterator.Builder add(RangeIterator range)
+        {
+            if (range == null)
+                return this;
+
+            if (range.getCount() > 0)
+            {
+                rangeIterators.add(range);
+                statistics.update(range);
+            }
+            else
+                FileUtils.closeQuietly(range);
+
+            return this;
+        }
+
+        public RangeIterator.Builder add(List<RangeIterator> ranges)
+        {
+            if (ranges == null || ranges.isEmpty())
+                return this;
+
+            ranges.forEach(this::add);
+            return this;
+        }
+
+        public int rangeCount()
+        {
+            return rangeIterators.size();
+        }
+
+        protected RangeIterator buildIterator()
+        {
+            switch (rangeCount())
+            {
+                case 1:
+                    return rangeIterators.get(0);
+
+                default:
+                    return new RangeUnionIterator(statistics, rangeIterators);
+            }
```

Comment:
The `switch` seems unnecessary for just two cases, we could use a simpler `if`  instead.